File: 1036-rotting-oranges/rotting-oranges.py
from collections import deque

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        fresh_count = 0
        rotten = deque()
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    fresh_count += 1
                elif grid[i][j] == 2:
                    rotten.append((i, j))
        
        if fresh_count == 0:
            return 0
        
        minutes = 0
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        
        while rotten:
            size = len(rotten)
            
            for _ in range(size):
                i, j = rotten.popleft()
                
                for di, dj in directions:
                    ni, nj = i + di, j + dj
                    
                    if 0 <= ni < m and 0 <= nj < n and grid[ni][nj] == 1:
                        grid[ni][nj] = 2
                        fresh_count -= 1
                        rotten.append((ni, nj))
            
            minutes += 1
        
        return minutes - 1 if fresh_count == 0 else -1

File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        
        queue<pair<int, int>> rotten;
        int freshCount = 0;
        
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 1) {
                    ++freshCount;
                } else if (grid[i][j] == 2) {
                    rotten.push({i, j});
                }
            }
        }
        
        if (freshCount == 0) return 0;
        
        int minutes = 0;
        vector<pair<int, int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        
        while (!rotten.empty()) {
            int size = rotten.size();
            
            for (int k = 0; k < size; ++k) {
                int i = rotten.front().first;
                int j = rotten.front().second;
                rotten.pop();
                
                for (const auto& dir : directions) {
                    int ni = i + dir.first;
                    int nj = j + dir.second;
                    
                    if (ni >= 0 && ni < m && nj >= 0 && nj < n && grid[ni][nj] == 1) {
                        grid[ni][nj] = 2;
                        --freshCount;
                        rotten.push({ni, nj});
                    }
                }
            }
            
            ++minutes;
        }
        
        return freshCount == 0 ? minutes - 1 : -1;
    }
};
